### PR TITLE
Bootstrap config

### PR DESCRIPTION
**Pull request by milan-zededa**
A new method for out-of-band delivery of the initial device configuration, which is sometimes needed to establish a working network connectivity between the device and the controller. Most of the design description and discussions is inside Zededa-internal documents, however the second commit of this PR contains documentation describing this feature at least within the scope of EVE OS.

Note that I had to make small adjustments in config processing by zedagent, domainmgr and nim, so that they would not be stuck waiting for device UUID but process the initial bootstrap config (just like a regular config from controller). In other words, EVE must have been enhanced to support config processing before onboarding.

### CODE REVIEW

File: docs/CONFIG.md [hunk 1]
```diff
@@ -5,29 +5,63 @@ While most of EVE's configuration is done via receiving a configuration object f
...
 
-Once an installer has been run on an edge node, the content of the CONFIG partition found on the installer media is simply copied into the CONFIG partition on the edge node's local storage. From that point on, even though these files reside in the read-write CONFIG partition on the edge node, they can NOT be changed by the controller. The only option for tweaking these files is to have debug console access either via local terminal, remote ssh terminal or batch removable media configuration.
+Once an installer has been run on an edge node, the content of the EVE partition found on the installer media is simply copied into the CONFIG partition on the edge node's local storage. From that point on, these files reside in the *read-only* CONFIG partition on the edge node and can NOT be changed by the controller.
```
Comment (eriknordmark) [inline docs/CONFIG.md:20]: Copy from CONFIG partition I assume.

File: docs/CONFIG.md [hunk 2]
```diff
@@ -5,29 +5,63 @@ While most of EVE's configuration is done via receiving a configuration object f
...
+* `configItems`
+* `systemAdapterList`
+* `deviceIoList`
+* `vlans`
+* `bonds`
+* `config_timestamp`
+
+Anything else will be silently ignored.
+
+The above allows to configure one or multiple management interfaces, as well as non-management interfaces, and can specify static IP and DNS configuration (for environments where DHCP is not used), plus the LTE configuration. In addition, it can specify proxies using several different mechanisms (explicitly, using WPAD, etc.). [Object level encryption](./OBJECT-LEVEL-ENCRYPTION.md) is not yet supported inside the bootstrap config, meaning that WiFi settings with the network password have to be left out it. Currently, it is therefore not possible to onboard device with WiFi-only connectivity using this mechanism.
```
Comment (eriknordmark) [inline docs/CONFIG.md:46]: "not yet" implies that WiFi will be added soon, but that isn't doable if we want the WiFi credentials to be encrypted; can't do that until the controller knows the device's ECDH certificate so it can encrypt for that target.
So I would strike "not yet" from the sentence.

File: pkg/pillar/zedcloud/authen.go [hunk 3]
```diff
@@ -422,36 +440,43 @@ func checkMimeProtoType(r *http.Response) bool {
 	return mimeType == ctTypeProtoStr
 }
 
-// VerifySigningCertChain - verify signing certificate chain from controller
-func VerifySigningCertChain(ctx *ZedCloudContext, content []byte) ([]byte, error) {
+// VerifyZControllerCert - unmarshal and verify the content of ZControllerCert
+// receiver from the controller.
+// Returns content of the signing certificate and the verification error/nil value.
+func VerifyZControllerCert(log *base.LogObject, content []byte) ([]byte, error) {
```
Comment (eriknordmark) [inline pkg/pillar/zedcloud/authen.go:446]: The "Z" in the name isn't informative.
How about naming this function VerifyProtoSigningCertChain or something which captures that it is protobuf decode plus VerifySigningCertChain?

File: pkg/pillar/cmd/nim/nim.go [hunk 4]
```diff
@@ -803,6 +799,12 @@ func (n *nim) handleZedAgentStatusImpl(_ string, statusArg interface{}) {
 // Otherwise the ShaFile and Shavalue is used to write the sha for the new file to avoid
 // re-application of the same config.
 func (n *nim) ingestDevicePortConfig() {
+	// Skip these legacy DPC json files if there is bootstrap config.
+	if _, err := os.Stat(types.BootstrapConfFileName); err == nil {
+		n.Log.Notice("Not ingesting DPC jsons from config partition: " +
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nim/nim.go:804]: This message makes one believe that there were some to ingest, which isn't necessarily the case.
How about moving down this block and prefix it with len(locations) != 0? That way the log message will indicate that something was ignored.

File: pkg/pillar/cmd/zedagent/handleconfig.go [hunk 5]
```diff
@@ -150,6 +152,104 @@ var nilUUID uuid.UUID
...
+
+	// Load file content.
+	contents, err := ioutil.ReadFile(types.BootstrapConfFileName)
+	if err != nil {
+		log.Errorf("Failed to read bootstrap config: %v", err)
+		return
+	}
+
+	// Mark bootstrap config as processed by storing SHA hash of its content
+	// under /persit/ingested.
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handleconfig.go:196]: typo

File: pkg/pillar/cmd/zedagent/handleconfig.go [hunk 6]
```diff
@@ -456,7 +555,7 @@ func getLatestConfig(url string, iteration int,
 	}
 	saveReceivedProtoMessage(contents)
 
-	return inhaleDeviceConfig(config, getconfigCtx, false)
+	return inhaleDeviceConfig(config, fromController, getconfigCtx)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handleconfig.go:558]: Can we make inhaleDeviceConfig and parseConfig have the same order of arguments? In your PR arg2 and 3 are different for the two functions.

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 7]
```diff
@@ -40,14 +40,36 @@ const (
...
+		getconfigCtx.lastReceivedConfig = time.Now()
+	}
+
+	// Make sure we do not accidentally revert to an older configuration.
+	// This depends on the controller attaching config timestamp.
+	// If not provided, the check is skipped.
+	if config.ConfigTimestamp.IsValid() {
+		configTimestamp := config.ConfigTimestamp.AsTime()
+		if getconfigCtx.lastConfigTimestamp.After(configTimestamp) {
+			log.Warnf("Skipping obsolete device configuration "+
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/parseconfig.go:65]: Can you add the config source to what is logged?

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 8]
```diff
@@ -40,14 +40,36 @@ const (
...
+	if getconfigCtx.localCommands != nil {
+		getconfigCtx.localCommands.Lock()
+		defer getconfigCtx.localCommands.Unlock()
+	}
+
+	// Update lastReceivedConfig even if the config processing is skipped.
+	if config.ConfigTimestamp.IsValid() {
+		getconfigCtx.lastReceivedConfig = config.ConfigTimestamp.AsTime()
+	} else {
+		getconfigCtx.lastReceivedConfig = time.Now()
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/parseconfig.go:56]: Odd to set lastReceivedConfig in the case where it is ignored due to being too old, since it would typically be overwritten (by the next getconfig). Move this block after the obsolete check?

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 9]
```diff
@@ -287,23 +287,32 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
...
 
 	initializeDirs()
 
-	// Apply saved radio config ASAP.
+	// Load bootstrap configuration if present.
 	getconfigCtx := zedagentCtx.getconfigCtx
+	loadBootstrapConfig(getconfigCtx)
+
+	// Get GlobalConfig.
+	// If not present (e.g. loading of bootstrap config failed), use default values.
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/zedagent.go:301]: How does this interact with upgradeconverter?

We need the logic in upgradeconverter which ensures that when a new item has been added in the new version of EVE-OS, it gets filled in with its default value.

Can't we have loadBootStrapConfig override all of configItemValueMap instead?

**Review by eriknordmark — APPROVED** [review]
Kicking off tests even though there are some comments to address.

File: pkg/pillar/cmd/zedagent/handleconfig.go [hunk 10]
```diff
@@ -150,6 +152,104 @@ var nilUUID uuid.UUID
...
+				log.Errorf("Failed to save SHA of bootstrap config: %v", err)
+			}
+		}
+	}()
+
+	// Unmarshal BootstrapConfig.
+	bootstrap := zconfig.BootstrapConfig{}
+	err = proto.Unmarshal(contents, &bootstrap)
+	if err != nil {
+		log.Errorf("Failed to unmarshal bootstrap config: %v", err)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handleconfig.go:212]: I wonder if we should make this and subsequent errors in this function result in some >10 LED blinking pattern.

**Comment by eriknordmark** [comment]
It seems like, appart from the testing effort, we could easily make this pb file format be used from a USB stick to repair the network configuration. That is possible because you made zedagent ignore anything but device network stuff and configItemValueMap, hence applications/volumes/network instances etc would not be accidentally deleted if a pb file is ingested from a USB stick. No need to tweak this PR but it might be useful follow-on work. Or we instead get the full protobuf config on a USB stick and skip the step of only repairing the network config?

**Review by eriknordmark — APPROVED** [review]
Run eden

File: docs/CONFIG.md (hunk 1 above)
Comment (milan-zededa) [reply docs/CONFIG.md:20]: I'm not sure, but when I prepare USB installer I see that the partition with configs is labeled as EVE.

File: docs/CONFIG.md (hunk 2 above)
Comment (milan-zededa) [reply docs/CONFIG.md]: Done

File: pkg/pillar/cmd/nim/nim.go (hunk 4 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/nim/nim.go]: Makes sense, done

File: pkg/pillar/cmd/zedagent/handleconfig.go (hunk 5 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/handleconfig.go]: fixed

File: pkg/pillar/cmd/zedagent/handleconfig.go (hunk 6 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/handleconfig.go]: OK, I have put context first, which is what we typically do. I have also fixed this for some other `parse*` methods in zedagent to make it unified.

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 7 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go:58]: done

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 8 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go]: true, moved the blocked after the skip

File: pkg/pillar/cmd/zedagent/handleconfig.go (hunk 10 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/handleconfig.go:225]: Done

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 9 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go:301]: Let me explain the current workflow:

1. upgradeconverter runs [importFromConfigPartition](https://github.com/milan-zededa/eve/blob/bootstrap-config/pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go#L80) (link to my fork with these changes included). If /`/config/bootstrap-config.pb` is NOT present, it will act as before, i.e. copying global.json from `/config` to `/persist/zedagent/ConfigItemValueMap/`. If bootstrap config is present, this step is skipped.
2. upgradeconverter runs [applyDefaultConfigItem](https://github.com/milan-zededa/eve/blob/bootstrap-config/pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go#L27). If global config is published (from previous boot), it just adds default values of missing keys (Eve was possibly upgraded and the new version has new config items). This is done regardless if there is any bootstrap config (i.e. the behavior is unchanged). If global config is missing (not published, likely the very first boot), then default values are published if bootstrap config is NOT present.
3. zedagent tries to load bootstrap config if present. If successful, config items (aka global config) are published. If not successful, zedagent publishes default values (with all keys known to the running EVE version).

So basically the goal was to prefer bootstrap config and publish default values only as a last resort. We could also make it so that default values are published by upgradeconverter and then possible get overwritten by zedagent during bootstrap loading, but there would be a small window during which the default config would be in effect, which could cause some confusion I think.
But in all those possible code-paths it is ensured that published global config has all the keys known to running EVE included, which was you main worry.

**Comment by milan-zededa** [comment]
@eriknordmark I have put your suggestions into a separate commit so that it is easier to review.
But somehow the integration tests were skipped (even before I have submitted requested changes)

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 9 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/zedagent.go:301]: In step 2 doesn't upgradeconverter always end up publishing the configItemValueMap? Or are you changing that?
I don't know what other parts of the system assume that there is a configItemValueMap...

File: docs/CONFIG.md (hunk 1 above)
Comment (eriknordmark) [reply docs/CONFIG.md:19]: You're correct; and the partition which has the rootfs has no label!

But to make this more clear, how about saying "the content of the partition with label EVE found on the installer media"?

File: pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go [hunk 11]
```diff
@@ -46,6 +47,11 @@ func applyDefaultConfigItem(ctxPtr *ucContext) error {
 			}
 		}
 	} else {
+		if bootstrapExists {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go:50]: Sorry, didn't see this earlier.

I don't know if other microservices will fail if they don't see a ConfigItemValueMap on startup, so I prefer not to change this part.
Can we instead have zedagent know that it should replace it in the case when it has a valid (newer timestamp etc) bootstrap config? That way we always have something in place for other agents hence less ordering dependencies to worry about.

File: pkg/pillar/cmd/zedagent/handleconfig.go [hunk 12]
```diff
@@ -150,6 +152,127 @@ var nilUUID uuid.UUID
...
+		return
+	}
+	err = proto.Unmarshal(payload, &devConfig)
+	if err != nil {
+		log.Errorf("Failed to unmarshal bootstrap config payload: %v", err)
+		return
+	}
+
+	// Apply bootstrap config.
+	inhaleDeviceConfig(getconfigCtx, &devConfig, fromBootstrap)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handleconfig.go:267]: Does this have the effect of publishing ConfigItemValueMap even if there are no such items in the config?
Or is the parseConfigItem* conditional on there being some items to process?

In any case, need to document these key side effects, and I think the function should have a different name like "maybeProcessBootstrapConfig()

File: pkg/pillar/cmd/zedagent/handleconfig.go [hunk 13]
```diff
@@ -150,6 +152,127 @@ var nilUUID uuid.UUID
...
+	log.Notice("Bootstrap config was applied")
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handleconfig.go:268]: This log entry will appear even if the bootstrap config is older than what was received from the controller.
Can we get some return value from inhale to indicate it was ignored and log things appropriately?

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 14]
```diff
@@ -39,15 +39,37 @@ const (
...
+	// Make sure we do not accidentally revert to an older configuration.
+	// This depends on the controller attaching config timestamp.
+	// If not provided, the check is skipped.
+	if config.ConfigTimestamp.IsValid() {
+		configTimestamp := config.ConfigTimestamp.AsTime()
+		if getconfigCtx.lastConfigTimestamp.After(configTimestamp) {
+			log.Warnf("Skipping obsolete device configuration "+
+				"(source: %v, timestamp: %v, currently applied: %v)",
+				source, configTimestamp, getconfigCtx.lastConfigTimestamp)
+			return false
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/parseconfig.go:61]: I don't know if returning true here (indicating it was skipped) is a good idea; skip due to old timestamp vs. skipped due to  in-progress reboot are two different things. Maybe introduce an enum for the return so the caller can tell the difference and log appropriately?

File: pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go (hunk 11 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go:50]: The main incentive was to prevent NIM from starting to apply lastresort config, which it does as soon as it sees global config. This has confused customers when they saw diag logs mentioning lastresort even though their override global.json has it disabled.
I checked (and tested) that all microservices start by waiting for global config:
```
	for !ctx.GCInitialized {
		log.Functionf("Waiting for GCInitialized")
		select {
		case change := <-subGlobalConfig.MsgChan():
			subGlobalConfig.ProcessChange(change)
		}
	}
	log.Functionf("processed GlobalConfig")
```
They do not start doing anything until it is available, therefore it is safe to delay generating global config until zedagent processes bootstrap config.

Comment (eriknordmark) [reply pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go:50]: OK, but I suspect there are comments about upgradecoverter and/or globalconfig/configitemvaluemap already existing. Such comments would need to be updated. Can you grep/search for such comments?

Comment (milan-zededa) [reply pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go:50]: I checked and the only related comments that keep repeating across microservices are:
`Pick up debug aka log level before we start real work`
`Look for global config such as log levels`
I would say they are both still OK.
It seems only zedagent had a mention of global config being guaranteed to exist, but that was removed by this PR:
```
-       // upgradeconverter ensures we have a ConfigItemValueMap so we
-       // read it to get the initial values
```

File: pkg/pillar/cmd/zedagent/handleconfig.go (hunk 12 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/handleconfig.go:274]: This is documented here: https://github.com/milan-zededa/eve/blob/bootstrap-config/pkg/pillar/cmd/zedagent/parseconfig.go#L2281-L2304
Basically zedagent merges default values with whatever is set in the received config. It then compares the new value with the previous config and publishes if a change was detected. I have made small improvement in the PR to ensure that zedagent tries to publish also when the previous config was not published successfully due to some (transient) error.
Also `loadBootstrapConfig` was renamed and suggested and comment was added to describe why it is conditional.

File: pkg/pillar/cmd/zedagent/handleconfig.go (hunk 13 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/handleconfig.go:275]: Added `configProcessingRetval` enum, now used by `getLatestConfig`, `inhaleDeviceConfig` and `parseConfig`.
Logging was corrected here.

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 14 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go:61]: Added `configProcessingRetval` enum, now used by `getLatestConfig`, `inhaleDeviceConfig` and `parseConfig`.

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 9 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go:302]: Resolved in the discussion below.

File: pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go (hunk 11 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go:50]: OK

**Review by eriknordmark — APPROVED** [review]
Re-run eden

**Comment by milan-zededa** [comment]
Checked the logs of failed eden tests but can't find anything relevant to this PR or to the failures themselves.
Also tried but couldn't reproduce these failures locally in the affected tests.

**Comment by milan-zededa** [comment]
@eriknordmark We had one integration test failing after retry with seemingly random failure. However, during E2E testing I found one small bug and pushed a fix for it (see the last commit), so I guess we are back to square one in terms of workflows.

**Review by eriknordmark — APPROVED** [review]
LGTM - once the workflows have completed the commits can be collapsed.
